filebeat/input/filestream/fswatch.go

```diff
@@ -137,9 +153,36 @@ func (w *fileWatcher) watch(ctx unison.Canceler) {
 			continue
 		}
 
+		// If we got notifications about harvesters being closed, update their
+		// file size accordingly.
+		//
+		// This is used to prevent a sort of race condition:
+		// When the reader/harvester reaches EOF, it blocks on a backoff,
+		// if during this time [logFile.shouldBeClosed] is called, marks the
+		// file as inactive and closes the reader context, once the backoff
+		// time expires the reader and harvester are closed without ingesting
+		// any more data.
+		//
+		// If the [fileWatcher] sends a write event while the harvester was blocked
+		// no new harvester is started because one is already running, however the
+		// [fileWatcher] updates its internal state and won't send write events until
+		// more data is added to the file.
+		//
+		// This can cause some lines to be missed because the harvester closed
+		// and the write event was lost.
+		//
+		// To prevent this from happening we get notified the offset of the file
+		// (data ingested) when the harvester closes. If we have this data we
+		// update our state to the same as the harvester, therefore starting
+		// a new harvester if needed.
+		if size, harvesterClosed := w.closedHarvesters[path]; harvesterClosed {
```

Comment:
Shouldn't this table use the file ID as the key? If a closed harvester was for a rotated file and we're now looking at a new one at the same path, we shouldn't apply the previous file's state.